### PR TITLE
feature: add support for typecasting

### PR DESCRIPTION
**Pull request by abidart**
*Issue #, if available:* 
Related to #10. 

*Description of changes:*
In an attempt to add support for int and float conversion, I made 3 changes:
- a typecast converter that converts int and float calls
- a typecast operator that adds a different behavior to int and float calls when they act on a qasm type
- a line in the transpiler file to incorporate the typecast transformer

The goal is to add a functionality such that the following two compile to the same QASM instructions, or very similar:

```
syndrome = int(measure([0,1]))
################################
a1 = aq.IntVar(measure(1))
a0 = aq.IntVar(measure(0))
syndrome = 2*a1 + a0
```

*Testing done:*
- I know that the current code fails when the call to `int` or `float` is nested, for example `syndrome = 2 * int(measure([0,1]))`. I need to fix that.
- The current changes pass `tox -e unit-tests`

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/amazon-braket/autoqasm/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/amazon-braket/autoqasm/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/autoqasm/blob/main/README.md) and [API docs](https://github.com/amazon-braket/autoqasm/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/autoqasm/transpiler/transpiler.py [hunk 1]
```diff
@@ -131,6 +137,7 @@ def transform_ast(
         node = assignments.transform(node, ctx)
         node = lists.transform(node, ctx)
         node = slices.transform(node, ctx)
+        node = typecast.transform(node, ctx)
```
Comment (rmshaffer) [inline src/autoqasm/transpiler/transpiler.py:140]: I think you'll want to call this _before_ the `assignments.transform` call. The reason is that you want any variable assignment to use the type _after_ it has been cast. This is so that `syndrome = int(measure[0,1])` produces an `int` variable in OpenQASM rather than a `bit`, as you're currently seeing.

File: src/autoqasm/operators/typecast.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    if aq_types.is_qasm_type(argument_):
+        if (
+            argument_.size is not None
+            and argument_.size > 1
+            and isinstance(argument_, aq_types.BitVar)
+        ):
+            typecasted_arg = type_to_aq_type_map[type_](argument_[0])
+            for index in range(1, argument_.size):
+                typecasted_arg += type_to_aq_type_map[type_](argument_[index]) * 2**index
+            return typecasted_arg
```
Comment (rmshaffer) [inline src/autoqasm/operators/typecast.py:43]: I am not sure that you need this logic here. OpenQASM spec [supports implicit casting](https://openqasm.com/language/types.html#allowed-casts) from `bit` to `int`. So if you simply create an `IntVar` and assign the `BitVar` to it, it should produce valid OpenQASM. The logic you have in the `else` block below might be enough.

File: src/autoqasm/converters/typecast.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            node (ast.stmt): AST node to transform.
+
+        Returns:
+            ast.stmt: Transformed node.
+        """
+        typecasts_supported = ["int", "float"]
+
+        template = """
+                ag__.typecast(type_, argument_)
+                """
```
Comment (rmshaffer) [inline src/autoqasm/converters/typecast.py:38]: Should you remove the newlines from this template? If you're just replacing a function call, the newlines might cause syntax errors in the transpiled Python.

**Review by rmshaffer — CHANGES_REQUESTED** [review]
Thanks for starting this! I've left a few comments/suggestions to help debug. Of course you'll want to add unit tests once the code is working, but that can be saved for a later iteration.

By the way, in case it's useful you can enable verbose autograph output via environment variable, e.g. you can do this in a Jupyter notebook cell:
`%env AUTOGRAPH_VERBOSITY 10` 
This will print out the intermediate transpiled program when you call `build()` on an AutoQASM program - I find this very useful when working on converters.

File: src/autoqasm/transpiler/transpiler.py (hunk 1 above)
Comment (abidart) [reply src/autoqasm/transpiler/transpiler.py]: That makes a lot of sense! I moved it up so that it gets called before assignments. I think that this is the main thing that was confusing me. I added a few changes in the other files to match this change.

File: src/autoqasm/converters/typecast.py (hunk 3 above)
Comment (abidart) [reply src/autoqasm/converters/typecast.py]: Done!

**Comment by abidart** [comment]
I added 3 unit tests but I was not sure if they should go into `test_api.py`, `test_operator.py`, or `test_converter.py`, what do you think? Also, do you think that the `expected_ir`s look reasonable?

File: test/unit_tests/autoqasm/test_api.py [hunk 4]
```diff
@@ -1266,3 +1266,39 @@ def test(int[32] a, int[32] b) {
...
+        @aq.main(num_qubits=2)
+        def main():
+            test = int(measure([0, 1]))
+
+        expected_ir = """OPENQASM 3.0;
+qubit[2] __qubits__;
+bit[2] __bit_0__ = "00";
+__bit_0__[0] = measure __qubits__[0];
+__bit_0__[1] = measure __qubits__[1];
+int[32] test = __bit_0__;"""
```
Comment (rmshaffer) [inline test/unit_tests/autoqasm/test_api.py:1282]: This one looks correct!

File: test/unit_tests/autoqasm/test_api.py [hunk 5]
```diff
@@ -1266,3 +1266,39 @@ def test(int[32] a, int[32] b) {
...
+        assert main.build().to_ir() == expected_ir
+
+    def test_int_typecasting_on_string(self):
+        @aq.main(num_qubits=2)
+        def main():
+            test = int("101", 2)
+
+        expected_ir = """OPENQASM 3.0;
+qubit[2] __qubits__;"""
```
Comment (rmshaffer) [inline test/unit_tests/autoqasm/test_api.py:1291]: This looks correct as well! (because the line `test = int("101", 2)` is pure Python, so it does not get captured in the program)

File: test/unit_tests/autoqasm/test_api.py [hunk 6]
```diff
@@ -1266,3 +1266,39 @@ def test(int[32] a, int[32] b) {
...
+    def test_nested_int_typecasting(self):
+        @aq.main(num_qubits=2)
+        def main():
+            test = 2 * int(measure([0, 1]))
+
+        expected_ir = """OPENQASM 3.0;
+qubit[2] __qubits__;
+bit[2] __bit_0__ = "00";
+__bit_0__[0] = measure __qubits__[0];
+__bit_0__[1] = measure __qubits__[1];"""
```
Comment (rmshaffer) [inline test/unit_tests/autoqasm/test_api.py:1303]: This doesn't look right. I would expect the output IR to contain something like:
```
int[32] __int_1__ = __bit_0__;
int[32] test = 2 * __int_1__;
```

File: src/autoqasm/operators/typecast.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+
+"""Operators for int cast statements."""
+
+from typing import Any
+
+from autoqasm import types as aq_types
+
+
+def int_typecast(argument_: Any, *args, **kwargs) -> aq_types.IntVar | int:
```
Comment (rmshaffer) [inline src/autoqasm/operators/typecast.py:22]: note that the Python 3.9 tests are failing due to the use of `|` on this line. To fix this, just add:
```
from __future__ import annotations
```
to the top of this file.

File: test/unit_tests/autoqasm/test_api.py (hunk 6 above)
Comment (rmshaffer) [reply test/unit_tests/autoqasm/test_api.py:1303]: For debugging, might be interesting to know what the IR output is if you modify the program to be like this:
```
        @aq.main(num_qubits=2)
        def main():
            test = int(2 * measure([0, 1]))
```
or like this:
```
        @aq.main(num_qubits=2)
        def main():
            test = 2 * int(measure([0, 1]))
            return test
```

File: test/unit_tests/autoqasm/test_api.py [hunk 8]
```diff
@@ -1266,3 +1266,39 @@ def test(int[32] a, int[32] b) {
...
+            test = int("101", 2)
+
+        expected_ir = """OPENQASM 3.0;
+qubit[2] __qubits__;"""
+        assert main.build().to_ir() == expected_ir
+
+    def test_nested_int_typecasting(self):
+        @aq.main(num_qubits=2)
+        def main():
+            test = 2 * int(measure([0, 1]))
```
Comment (rmshaffer) [inline test/unit_tests/autoqasm/test_api.py:1297]: to fix linter error:
```suggestion
            test = 2 * int(measure([0, 1]))  # noqa: F841
```

File: test/unit_tests/autoqasm/test_api.py [hunk 9]
```diff
@@ -1266,3 +1266,39 @@ def test(int[32] a, int[32] b) {
...
+bit[2] __bit_0__ = "00";
+__bit_0__[0] = measure __qubits__[0];
+__bit_0__[1] = measure __qubits__[1];
+int[32] test = __bit_0__;"""
+        assert main.build().to_ir() == expected_ir
+
+    def test_int_typecasting_on_string(self):
+        @aq.main(num_qubits=2)
+        def main():
+            test = int("101", 2)
```
Comment (rmshaffer) [inline test/unit_tests/autoqasm/test_api.py:1288]: to fix linter error:
```suggestion
            test = int("101", 2)  # noqa: F841
```

File: test/unit_tests/autoqasm/test_api.py [hunk 10]
```diff
@@ -1266,3 +1266,39 @@ def test(int[32] a, int[32] b) {
 test(2, 3);
 test(4, 5);"""
     assert main.build().to_ir() == expected
+
+
+class TestTypecasting:
+    def test_int_typecasting_on_measure(self):
+        @aq.main(num_qubits=2)
+        def main():
+            test = int(measure([0, 1]))
```
Comment (rmshaffer) [inline test/unit_tests/autoqasm/test_api.py:1275]: to fix linter error:
```suggestion
            test = int(measure([0, 1]))  # noqa: F841
```

**Review by rmshaffer — CHANGES_REQUESTED** [review]
This looks like it's on the right track! A few more comments/suggestions here.

File: test/unit_tests/autoqasm/test_api.py (hunk 6 above)
Comment (abidart) [reply test/unit_tests/autoqasm/test_api.py]: I checked and using
```
@aq.main(num_qubits=2)
        def main():
            test = 2 * IntVar(measure([0, 1]))
```
results in the same IR. I think the discrepancy came from the fact that the variable `test` was not being used after declared. Your second suggestion helped me realize that adding `return test` or other lines that make use of `test` results in an IR with the int[32] lines. I update the unit test to reflect this.

On a second note, I did not add support for arithmetic operations on BitVara in this PR because I focused on int typecasting. To add the arithmetic operations, would you suggest I add methods like `__mul__` and `__rmul__` to the `BitVar` class (where `self` gets casted to an `IntVar`) or use a converter to intercept the arithmetic operations and cast the BitVars to IntVars?

**Review by rmshaffer — CHANGES_REQUESTED** [review]

**Comment by rmshaffer** [comment]
> I added 3 unit tests but I was not sure if they should go into `test_api.py`, `test_operator.py`, or `test_converter.py`, what do you think?

@abidart Probably `test_operators.py` makes the most sense. `test_api.py` is getting too large and needs to be split apart at some point.

File: src/autoqasm/operators/typecast.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    already declared.
+
+    Args:
+        argument_ (Any): object to be converted into an int.
+
+    Returns:
+        IntVar | int : IntVar object if argument is QASM type, else int.
+    """
+    if aq_types.is_qasm_type(argument_):
+        return aq_types.IntVar(argument_)
```
[the remark this answers is not in the document]
Comment (abidart) [reply src/autoqasm/operators/typecast.py:34]: Super clear! Thank you so much.

I wanted to double-check one last thing. If we consider the test:

```
@aq.main(num_qubits=2)
        def main():
            test = int(measure([0, 1]))
```
I am now getting the following instructions:
```
int[32] test;
qubit[2] __qubits__;
bit[2] __bit_0__ = "00";
__bit_0__[0] = measure __qubits__[0];
__bit_0__[1] = measure __qubits__[1];
int[32] __int_1__;
__int_1__ = __bit_0__;
test = __int_1__;
```
I am particularly interested in the last 3 lines. Would you have expected something like `test = __bit_0__` (given that `test` was already declared to be an `int` in the first line) or is declaring `__int_1__` and setting it equal to `__bit_0__` something expected due to the conversion?

[the remark this answers is not in the document]
Comment (rmshaffer) [reply src/autoqasm/operators/typecast.py:34]: Yes, this is completely expected because of the way the conversion is happening. You're creating an IntVar (`__int_1__`) from the typecast, and then the assignment is creating another IntVar (`test`).

In the future we could certainly look for ways to optimize the generated IR, but right now we don't do that.

File: test/unit_tests/autoqasm/test_api.py [hunk 12]
```diff
@@ -1266,3 +1266,43 @@ def test(int[32] a, int[32] b) {
...
+
+        expected_ir = """OPENQASM 3.0;
+qubit[2] __qubits__;"""
+        assert main.build().to_ir() == expected_ir
+
+    def test_nested_int_typecasting(self):
+        @aq.main(num_qubits=2)
+        def main():
+            test = 2 * int(measure([0, 1]))  # noqa: F841
+            return test
```
[the remark this answers is not in the document]
Comment (abidart) [reply test/unit_tests/autoqasm/test_api.py]: Added! Could you confirm that in the case where there is no return statement, there should not be an instruction involving the ` 2 *`. Currently, I am getting:

*with* return
```
OPENQASM 3.0;
output int[32] test;
qubit[2] __qubits__;
bit[2] __bit_0__ = "00";
__bit_0__[0] = measure __qubits__[0];
__bit_0__[1] = measure __qubits__[1];
int[32] __int_1__;
__int_1__ = __bit_0__;
test = 2 * __int_1__;
```

*without* return
```
OPENQASM 3.0;
qubit[2] __qubits__;
bit[2] __bit_0__ = "00";
__bit_0__[0] = measure __qubits__[0];
__bit_0__[1] = measure __qubits__[1];
int[32] __int_1__;
__int_1__ = __bit_0__;
```

[the remark this answers is not in the document]
Comment (rmshaffer) [reply test/unit_tests/autoqasm/test_api.py]: That's interesting - I still would expect the `test` variable to be declared and assigned `2 * __int_1__` there. But functionally it doesn't matter, of course. I expect this would be an issue in the assignment converter and outside the scope of your PR. It shouldn't block moving forward with your changes.

[the remark this answers is not in the document]
Comment (abidart) [reply test/unit_tests/autoqasm/test_api.py]: Maybe I can look into this in a new PR?

**Review by rmshaffer — DISMISSED** [review]

**Comment by abidart** [comment]
@rmshaffer I pushed a new commit with cosmetic changes based on the review from #29 anticipating similar comments.

File: test/unit_tests/autoqasm/test_operators.py [hunk 13]
```diff
@@ -932,6 +932,65 @@ def test_list_ops():
...
+
+    expected_ir = """OPENQASM 3.0;
+output int[32] test;
+qubit[2] __qubits__;
+bit[2] __bit_0__ = "00";
+__bit_0__[0] = measure __qubits__[0];
+__bit_0__[1] = measure __qubits__[1];
+int[32] __int_1__;
+__int_1__ = __bit_0__;
+test = 2 * __int_1__;"""
```
Comment (rmshaffer) [inline test/unit_tests/autoqasm/test_operators.py:991]:
```suggestion
test = 2 * __int_1__;"""
    assert main.build().to_ir() == expected_ir
```

Comment (rmshaffer) [reply test/unit_tests/autoqasm/test_operators.py:991]: Just fixing my own bad merge here when I merged the latest `main`, sorry about that @abidart.

File: test/unit_tests/autoqasm/test_operators.py [hunk 14]
```diff
@@ -932,6 +932,65 @@ def test_list_ops():
...
+
+    expected_ir = """OPENQASM 3.0;
+qubit[2] __qubits__;"""
+    assert main.build().to_ir() == expected_ir
+
+
+def test_nested_int_typecasting_without_return():
+    @aq.main(num_qubits=2)
+    def main():
+        test = 2 * int(measure([0, 1]))  # noqa: F841
```
[the remark this answers is not in the document]
Comment (rmshaffer) [reply test/unit_tests/autoqasm/test_operators.py:965]: Some discussion of that here: https://github.com/amazon-braket/autoqasm/pull/27#discussion_r1633320359

tl;dr: This also looks like a bug to me, but I think it's more likely something pre-existing in the `assignments` converter than anything related to this PR.

[the remark this answers is not in the document]
Comment (abidart) [reply test/unit_tests/autoqasm/test_operators.py:965]: How about I replace the test without the return statement with the following?

```
@pytest.mark.xfail(reason="Bug: assignments do not work as expected when operators are nested")
def test_nested_int_typecasting_without_return():
    @aq.main(num_qubits=2)
    def main():
        test = 2 * int(measure([0, 1]))  # noqa: F841

    expected_ir = """OPENQASM 3.0;
qubit[2] __qubits__;
bit[2] __bit_0__ = "00";
__bit_0__[0] = measure __qubits__[0];
__bit_0__[1] = measure __qubits__[1];
int[32] __int_1__;
__int_1__ = __bit_0__;
test = 2 * __int_1__;"""
    assert main.build().to_ir() == expected_ir
```

File: test/unit_tests/autoqasm/test_operators.py [hunk 15]
```diff
@@ -932,6 +932,66 @@ def test_list_ops():
...
+int[32] __int_1__;
+__int_1__ = __bit_0__;
+test = __int_1__;"""
+    assert main.build().to_ir() == expected_ir
+
+
+def test_int_typecasting_on_string():
+    @aq.main(num_qubits=2)
+    def main():
+        test = int("101", 2)  # noqa: F841
```
Comment (laurencap) [inline test/unit_tests/autoqasm/test_operators.py:955]: Should we raise an error, instead?

Comment (abidart) [reply test/unit_tests/autoqasm/test_operators.py:955]: Hello @laurencap, thank you so much for the review! My goal with this test was to have a "normal" call to `int` within a program to check that the built-in `int` function gets called—correctly, with all passed arguments—whenever the first input does not involve a QASM type. Could you explain me how can I change it?

**Review by rmshaffer — APPROVED** [review]

File: test/unit_tests/autoqasm/test_operators.py (hunk 15 above)
Comment (rmshaffer) [reply test/unit_tests/autoqasm/test_operators.py:955]: I believe this is working as intended, but maybe the confusion could be cleared up by changing the name of the test function to make it more clear what is being tested - e.g. `test_int_typecasting_on_python_string_not_captured`

**Review by rmshaffer — APPROVED** [review]
